---
title: "multi_omics_visualization"
author: 
- "ddedesener"
- "DeniseSl22"
date: "03/05/22"
output:
 md_document:
    variant: markdown_github
always_allow_html: true
---
## Introduction
In this script, we visualize multi-omics data in Cytoscape through the WikiPathways App for Cytoscape.
Through this app, pathways from WikiPathways, Reactome and LIPID MAPS are available.
Identifiers from these three pathway model databases are unified and harmonized to Ensembl and ChEBI IDs, to allow for multi-omics data visualization. 
The transcriptomics data and metabolites data are combined to automatically visualize their respective log2FC and p-values. 

## Setup
```{r setup, warning=FALSE, message=FALSE}
# check if libraries are already installed > otherwise install it
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if(!"rstudioapi" %in% installed.packages()) BiocManager::install("rstudioapi")
if(!"RCy3" %in% installed.packages()) BiocManager::install("RCy3")
if(!"rWikiPathways" %in% installed.packages()) BiocManager::install("rWikiPathways")
if(!"RColorBrewer" %in% installed.packages()) BiocManager::install("RColorBrewer")
if(!"dplyr" %in% installed.packages()) BiocManager::install("dplyr")
#load libraries
library(rstudioapi)
library(RCy3)#connect cytoscape via R
library(rWikiPathways)#to get pathways from WikiPathways
library(RColorBrewer)#to manage colors with R
library(dplyr)
# set your working environment to the location where your current source file is saved into.
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
```

## Import data from step 1 and 3
```{r datasets_import, warning = FALSE, message = FALSE}
##TODO: start workflow on data in file from this folder (in case step 1 and 3 are not run yet?)

##Transcriptomics
tSet_CD <- read.delim("../1-identifier_mapping_transcriptomics/results/IDMapping_CD")
tSet_UC <- read.delim("../1-identifier_mapping_transcriptomics/results/IDMapping_UC")

##Take a subset of the transcriptomics data, which is relevant for the multiomics-visualization:
tSet_CD_ileum_multi <- tSet_CD[,c('Ensembl.ID_PriID_BridgeDb', 'log2FC_ileum', 'pvalue_ileum')]
tSet_CD_rectum_multi <- tSet_CD[,c('Ensembl.ID_PriID_BridgeDb', 'log2FC_rectum', 'pvalue_rectum')]
tSet_UC_ileum_multi <- tSet_UC[,c('Ensembl.ID_PriID_BridgeDb', 'log2FC_ileum', 'pvalue_ileum')]
tSet_UC_rectum_multi <- tSet_UC[,c('Ensembl.ID_PriID_BridgeDb', 'log2FC_rectum', 'pvalue_rectum')]

##Rename the columns to be able to merge the data with metabolomics later:
dfs <- c("tSet_CD_ileum_multi", "tSet_CD_rectum_multi", "tSet_UC_ileum_multi", "tSet_UC_rectum_multi")
for(df in dfs)
  assign(df, setNames(get(df),  c("Identifier","log2FC","pvalue")))

##Metabolomics
mSet_CD <- read.delim("../3-identifier_mapping_metabolomics/results/mbx_IDMapping_CD")
mSet_UC <- read.delim("../3-identifier_mapping_metabolomics/results/mbx_IDMapping_UC")

##Take a subset of the metabolomics data, which is relevant for the multiomics-visualization:
mSet_CD_multi <- mSet_CD[,c('ChEBI_PriID_BridgeDb', 'log2FC', 'pvalue')]
mSet_UC_multi <- mSet_UC[,c('ChEBI_PriID_BridgeDb', 'log2FC', 'pvalue')]

##Rename the columns to be able to merge the data with metabolomics later:
dfs <- c("mSet_CD_multi", "mSet_UC_multi")
for(df in dfs)
  assign(df, setNames(get(df),  c("Identifier","log2FC","pvalue")))

##Merge two OMICS types together:
combined_CD_ileum_multi <- rbind(tSet_CD_ileum_multi,mSet_CD_multi)
combined_CD_rectum_multi <- rbind(tSet_CD_rectum_multi,mSet_CD_multi)
combined_UC_ileum_multi <- rbind(tSet_UC_ileum_multi,mSet_UC_multi)
combined_UC_rectum_multi <- rbind(tSet_UC_rectum_multi,mSet_UC_multi)

##Remove NA values for missing IDs.
dfs <- c("combined_CD_ileum_multi", "combined_CD_rectum_multi", "combined_UC_ileum_multi", "combined_UC_rectum_multi")
for(df in dfs)
  assign(df, na.omit(get(df)))

rm(list = setdiff(ls(), dfs)) # removing variables that are not required
names_of_dataframes <- ls.str(mode = "list")
dfs <- sapply(ls(), get)
```

## Select relevant pathway based on overlap
```{r}
##TODO: to be added; based on overlap visualization @Tooba?
```

## Import pathway
```{r import, warning=FALSE, message=FALSE}
#make sure to first launch Cytoscape outside of Rstudio, v.3.9.1
cytoscapePing()
##check metadata of tool
cytoscapeVersionInfo ()
#close all opened session before starting
closeSession(FALSE)
##Set up WikiPathways app in Cytoscape, v.3.3.10
if("WikiPathways" %in% commandsHelp("")) print("Success: the WikiPathways app is installed") else print("Warning: WikiPathways app is not installed. Please install the WikiPathways app before proceeding.")
if(!"WikiPathways" %in% commandsHelp("")){
  installApp("WikiPathways")
}
#pathway IDs to be visualized
pathway.id <- "WP4726"# Sphingolipid metabolism: integrated pathway 
##Select a dataset to visualize on the pathway:
i <- 2 #(options are 1,  2, 3 and 4)
#import pathways as pathway in cytoscape
RCy3::commandsRun(paste0('wikipathways import-as-pathway id=',pathway.id )) 
```

## Data upload
```{r data_upload, warning=FALSE, message=FALSE}
#get node table from imported pathway in cytoscape
ID.cols <- getTableColumns(table ="node", columns = c("XrefId","Ensembl", "ChEBI"))
#filter out rows which contain NA value for columns Ensembl and ChEBI
ID.cols <- ID.cols[!with(ID.cols, is.na(Ensembl) & is.na(ChEBI)),]
#if a row value in the Ensembl column is NA then replace it with ChEBI  
ID.cols$Ensembl <- ifelse(is.na(ID.cols$Ensembl), ID.cols$ChEBI, ID.cols$Ensembl)
#use the only one column contains both Ensembl and ChEBI identifiers
ID.cols <- data.frame(ID.cols[,c(1,2)])
#change column name
colnames(ID.cols)[2] <- "omics.ID"
#merge two data frames for adding xrefid to the combined data frame
data <- merge(dfs[[i]], ID.cols, by.x = "Identifier", by.y = "omics.ID" )
#remove duplicate rows
data <- data %>% distinct(Identifier, .keep_all = TRUE)
#change column order
data <- data [,c(1,2,3,4)]
colnames(data)[2] <- "log2FC"
#load data to the imported pathway in cytoscape by key column as XrefId
loadTableData(table = "node", data = data, data.key.column = "XrefId", table.key.column = "XrefId")
```
## Visualization options
```{r visualization, warning=FALSE, message=FALSE}
#new visual style is created
RCy3::copyVisualStyle("default","pathwayStyle")
#set new style as the current style
RCy3::setVisualStyle("pathwayStyle")
#set node dimensions as fixed sizes
RCy3::lockNodeDimensions(TRUE, style.name="pathwayStyle")

#node shape mapping
RCy3::setNodeShapeMapping('Type',c('GeneProduct','Protein', 'Metabolite'),c('ELLIPSE','ELLIPSE','RECTANGLE'),style.name="pathwayStyle")
#change node height
RCy3::setNodeHeightMapping('Type',c('GeneProduct','Protein', 'Metabolite'), c(23,23,25), mapping.type = "d", style.name = "pathwayStyle")
#change node width
RCy3::setNodeWidthMapping('Type',c('GeneProduct','Protein', 'Metabolite'), c(60,60,175), mapping.type = "d", style.name = "pathwayStyle")

#set node color based on log2FC for both genes and metabolites
node.colors <- c(rev(brewer.pal(3, "RdBu")))
setNodeColorMapping("log2FC", c(-1,0,1), node.colors, default.color = "#D3D3D3", style.name = "pathwayStyle")

#set node border width and color based on p-value
#first we need to get all p-values from node table
pvalues  = getTableColumns(table = 'node',columns = c('name','pvalue'))  
min.pval = min(pvalues[,2],na.rm=TRUE)
#get non-significant nodes (p-value>0.05)
nonsign.nodes <- pvalues  %>% filter(pvalue > 0.05, na.rm = TRUE)
#filter out duplicated rows
nonsign.nodes <- unique(nonsign.nodes)

#set node border width for all nodes which has a p-value
setNodeBorderWidthMapping('pvalue', c(min.pval,0.05), c(6,6) , mapping.type = "c", style.name = "pathwayStyle")
#set node border color for all nodes which has a p-value
setNodeBorderColorMapping('pvalue', c(min.pval,0.05), c('#00C400','#00C400'), style.name = "pathwayStyle")

#filter out non-sign nodes from the visualization
RCy3::setNodeBorderWidthBypass(nonsign.nodes$name, new.sizes = 2)
RCy3::setNodeBorderColorBypass(nonsign.nodes$name, new.colors = "#D3D3D3")

#save output 
filename_multiomics <- paste0("results/", pathway.id, "_", names_of_dataframes[i], "omics_visualization.png")
png.file <- file.path(getwd(), filename_multiomics)
exportImage(png.file,'PNG', zoom = 500)
```
