---
title: "Identifier mapping (transcriptomics)"
author: 
- "DeniseSl22"
- "tabbassidaloii"
- "ddedesener"
date: "20/06/2022"
output:
 md_document:
    variant: markdown_github
always_allow_html: true
---
## Introduction
In this section, identifier (IDs) mapping is performed on an example transcriptomics data set, which was original annotated using HGNC symbols.
The dataset has been preprocessed already, for details see step 1 and 2 of the multi-omics workflow at: https://github.com/BiGCAT-UM/Transcriptomics_Metabolomics_Analysis/tree/master/transcriptomics_analysis .
We map the HGNC symbols to Entrez Gene and Ensembl IDs, since tools downstream of this step require different input formats for the IDs.

We use two tools for this mapping; first we use org.Hs.eg.db [doi:10.18129/B9.bioc.org.Hs.eg.db] and AnnotationDbi; second we use BridgeDb [doi:10.18129/B9.bioc.BridgeDbR].

## R environment setup
```{r setup, warning=FALSE, message=FALSE}
# check if libraries are already installed, otherwise install it
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if(!"rstudioapi" %in% installed.packages()) BiocManager::install("rstudioapi")
if(!"org.Hs.eg.db" %in% installed.packages()) BiocManager::install("org.Hs.eg.db")  
if(!"AnnotationDbi" %in% installed.packages()) BiocManager::install("AnnotationDbi")
if(!"dplyr" %in% installed.packages()) install.packages("dplyr")

#load installed libraries
library(rstudioapi) # interface for interacting with RStudio IDE with R code.
library(org.Hs.eg.db) #This is the organism annotation package ("org") for Homo sapiens ("Hs"), organized as an AnnotationDbi   package ("db"), using Entrez Gene IDs ("eg") as primary key.
library(AnnotationDbi) # for connecting and querying annotation databases
library(dplyr)

# set your working environment to the location where your current source file is saved into.
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
```

## Importing dataset
The data will be read for the disease on two biopsy locations
```{r dataset, warning=FALSE, message=FALSE}
#we have four datasets, two different disorders and two different biopsy locations
#Reading the files for UC disorder
(filenames <- list.files("data", pattern = "UC", full.names = TRUE)) #Printing the file names
dataset_UC <- lapply(filenames, read.delim) #Reading the files

#filter out  unused columns, we select geneSymbol, log2FC and pvalue and
#merge two dataset of two locations into one datafile per disorder
dataset_UC <- merge(dataset_UC[[1]] %>% select ("X", "FoldChange", "padj"), #Merging and sub-setting the columns required
                    dataset_UC[[2]] %>% select ("X", "FoldChange", "padj"),
                    by = "X", all = TRUE)

#Reading the files for CD disorder
(filenames <- list.files("data", pattern = "CD", full.names = TRUE)) #Printing the file names
dataset_CD <- lapply(filenames, read.delim) #Reading the files

#filter out  unused columns, we select geneSymbol, log2FC and pvalue and
#merge two dataset of two locations into one datafile
dataset_CD <- merge(dataset_CD[[1]] %>% select ("X", "FoldChange", "padj"), #Merging and sub-setting the columns required
                    dataset_CD[[2]] %>% select ("X", "FoldChange", "padj"),
                    by = "X", all = TRUE)

#change column names
colnames(dataset_CD) <- colnames(dataset_UC) <- c("GeneSymbol","log2FC_ileum","pvalue_ileum","log2FC_rectum","pvalue_rectum")

rm(list = setdiff(ls(), c("dataset_UC", "dataset_CD"))) # removing variables that are not required
```

## Converting hgnc gene symbols to the corresponding Entrez (NCBI) gene IDs
```{r converting_entrez,warning=FALSE, message=FALSE}
#converting gene symbols to entrez ID since these are required for the enrichR function
hs <- org.Hs.eg.db #This object is a simple mapping of Entrez Gene identifier

# check if all the gene symbols are the same in both datasets
all(dataset_CD$GeneSymbol == dataset_UC$GeneSymbol) 
##same gene symbols, so mapping based on one of the files 
entrezID <- AnnotationDbi::select(hs, keys = dataset_CD$GeneSymbol,
            columns = c("ENTREZID", "SYMBOL"),
            keytype = "SYMBOL")

#store double mapping info
entrezID_doubles <- (length((na.omit(entrezID$SYMBOL))) - length(unique(na.omit(entrezID$SYMBOL))))
# entrezID_doubles <- names (table(entrezID$SYMBOL)[table(entrezID$SYMBOL) > 1])
# entrezID [entrezID$SYMBOL %in% entrezID_doubles,] # If you want to see which genes have duplicated Entrez Gene identifier

#filter out double gene symbols
entrezID <- entrezID %>% distinct (entrezID$SYMBOL, .keep_all = TRUE)

# add entrezIDs for each gene symbol in the dataset
dataset_CD$ENTREZ.ID_org.Hs <- entrezID$ENTREZID [match (dataset_CD$GeneSymbol, entrezID$SYMBOL)] 
dataset_UC$ENTREZ.ID_org.Hs <- entrezID$ENTREZID [match (dataset_UC$GeneSymbol, entrezID$SYMBOL)] 

rm(list = setdiff(ls(), c("dataset_UC", "dataset_CD", "entrezID_doubles", "hs"))) # removing variables that are not required
```

## Converting hgnc gene symbols to the corresponding Ensembl IDs
```{r converting_ensembl,warning=FALSE, message=FALSE}
#converting gene symbols to Ensembl ID since these are required for the Cytoscape multiomics visualization
ensemblID <- AnnotationDbi::select(hs, keys = dataset_CD$GeneSymbol,
            columns = c("ENSEMBL", "SYMBOL"),
            keytype = "SYMBOL")
#store double mapping info
ensemblID_doubles <- (length((na.omit(ensemblID$SYMBOL))) - length(unique(na.omit(ensemblID$SYMBOL))))
#filter out double gene symbols
ensemblID <- ensemblID %>% distinct (ensemblID$SYMBOL, .keep_all = TRUE)

##CD
# add entrezIDs for each gene symbol in the dataset
dataset_CD <- cbind(ensemblID$ENSEMBL,dataset_CD)
#change column name
colnames(dataset_CD)[1] = "Ensembl.ID_org.Hs"
##UC
# add entrezIDs for each gene symbol in the dataset
dataset_UC <- cbind(ensemblID$ENSEMBL,dataset_UC)
#change column name
colnames(dataset_UC)[1] = "Ensembl.ID_org.Hs"
```

##Mapping stats:
```{r mapping stats,warning=FALSE, message=FALSE}
##HGNC_originalData
paste0("The total number of datapoints for the transcriptomics dataset is: ", nrow(dataset_CD))
paste0("The total number of unique HGNC Symbol IDs for the transcriptomics dataset is: ", length(unique(na.omit(dataset_CD$GeneSymbol))))
paste0("~~~~~")
##Entrez_org.Hs
paste0("The total number of missing mappings for HGNC Symbol to Entrez IDs from the org.Hs mapping is: ", sum(is.na(dataset_CD$ENTREZ.ID_org.Hs)))
paste0("The total number of Entrez IDs from the org.Hs mapping is: ", (nrow(dataset_CD)-sum(is.na(dataset_CD$ENTREZ.ID_org.Hs))))
paste0("The total number of unique Entrez IDs from the org.Hs mapping is: ", length(unique(na.omit(dataset_CD$ENTREZ.ID_org.Hs))))
paste0("The total number of double (one to many) mappings for Entrez IDs from the org.Hs mapping is: ", entrezID_doubles)
paste0("~~~~~")
##Ensembl_org.Hs
paste0("The total number of missing mappings for HGNC Symbol to Ensembl IDs from the org.Hs mapping is: ", sum(is.na(dataset_CD$Ensembl.ID_org.Hs)))
paste0("The total number of Ensembl IDs from the org.Hs mapping is: ", (nrow(dataset_CD)-sum(is.na(dataset_CD$Ensembl.ID_org.Hs))))
paste0("The total number of unique Ensembl IDs from the org.Hs mapping is: ", length(unique(na.omit(dataset_CD$Ensembl.ID_org.Hs))))
paste0("The total number of double (one to many) mappings for Ensembl IDs from the org.Hs mapping is: ", ensemblID_doubles)

##Citation org.Hs:
citation("org.Hs.eg.db")

##Citation BridgeDb:
citation("BridgeDbR")
```

##Save data, print session info and remove large datasets:
```{r print_session_info}
##Save data: exporting results to the file
#CD
write.table(dataset_CD, file="results/IDMapping_CD",
            sep = "\t" ,quote = FALSE, row.names = FALSE)

#UC
write.table(dataset_UC, file="results/IDMapping_UC",
            sep = "\t" ,quote = FALSE, row.names = FALSE)

##Print session info:
sessionInfo()

##Remove data objects which are not needed for further processing:
rm(list=setdiff(ls(), c("dataset_CD", "dataset_UC", "work_DIR")))
```
